### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by dan1dr**
**Table of contents**

1. **push_exploration_non_linear**: contains the EDA
2. **model_fit.py**: structured code for model fitting
3. **model_predict.py**: structured code for returning the predictions

image

- **push_exploration_non_linear.ipynb**: quick code for playing with models, some exploration and experimentation. Again, do not feel that every part needs to be reviewed. I noticed that you pointed out in the previous module several times that having two PR curves doesn't provide meaningful insights without defining a working area, and I understand that. In addition, I'd like to emphasize that I have a rough idea of what my working area might be, but I haven't defined a fixed point yet. My decision will depend on the model's performance. For example, if my logistic regression performs exceptionally well at a point with low recall, while my CatBoost model performs poorly there but excels at higher recalls with sufficient precision, this could influence my decision-making. There may not be a fixed threshold for certain parameters in this case (assuming both models meet the expectation of having probabilities above 0.05 in both interesting areas and any of those would be a better approach vs. baseline). Please let me know if this makes sense to you. In this case, I chose to focus on my orange square point based on some conditions described at the end of the notebook, assuming it could be open for discussion and even we could define some costs to each uninstallation, purchase, etc.

- **model_fit.py**: clean and well-structured code for fitting the model and returning the required dictionary. I've included a function _called find_precision_threshold_ before training the model to return the corresponding probability, which is then used in _model_predict.py_. Please feel free to discuss this approach. I understand that a working point can be defined with various thresholds (for example, in Pablo's example, it was the point given a probability of 0.05). Additionally, please provide comments on how the dictionary _event_ is used at the end of the code and if it is passed correctly within the _handler_fit_ function.

- **model_predict.py**: clean and structured code for making predictions. Defined 3 examples to doublecheck it was indeed returning binary predictions.

### Comments

1. I haven't implemented any unit tests to catch up with current assignments (feel free to propose any potential tests different from the ones made in PR3).
2. I have incorporated some minor comments from PR3 into clean functions to enhance clarity and eliminate unnecessary returns.

Thank you for the feedback!

### CODE REVIEW

File: src/module_4/push_exploration_non_linear.md [hunk 1]
```diff
@@ -0,0 +1,2005 @@
...
+def read_data(file_path: str) -> pd.DataFrame:
+    """
+    Read and validate data from a CSV file.
+    Parameters: file_path (str)
+    Returns: pd.DataFrame
+    """
+    try:
+        data = pd.read_csv(file_path, index_col=False)
+
+        if data is not None and not data.empty:
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:67]: Small comment: I don't know if this will if statement will be ever be False, since you are already capturing: `pd.errors.EmptyDataError`

File: src/module_4/push_exploration_non_linear.md [hunk 2]
```diff
@@ -0,0 +1,2005 @@
...
+def preprocess_data(
+    data: pd.DataFrame, remove_if_all_na: bool = False, num_items: int = 5
+) -> pd.DataFrame:
+    """
+    Preprocess data by removing rows with
+    filter orders with at least 5 items
+
+    Parameters:
+    - data (pd.DataFrame)
+    - remove_all_na_rows (bool): If True, remove rows where at least
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:104]: The name of this variable is quite misleading. I would expect that `remove_all_na_rows` being `True`, removes a row if ALL the values are missing (in your description this happens when the argument is `False`)
Apart from it, if a row has ALL values missing it is not useful so I'd remove it by default and therefore I am not sure if this parameter makes a lot of sense.

File: src/module_4/push_exploration_non_linear.md [hunk 3]
```diff
@@ -0,0 +1,2005 @@
...
+    1 value is missing. If False, remove rows where all values are missing.
+
+    Returns:
+    - pd.DataFrame. The preprocessed dataset
+    """
+    try:
+        initial_length = len(data)
+        if remove_if_all_na:  # remove if everything is NA
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:112]: Ah okey, here you are doing the opposite that you said in the comments. However, as said before, I'd always delete rows with ALL nas and have a variable `drop_rows_with_some_na` which drops columns with at least 1 NA or not

File: src/module_4/push_exploration_non_linear.md [hunk 4]
```diff
@@ -0,0 +1,2005 @@
...
+        processed_data = data[num_items_ordered >= num_items]
+
+        logging.info(f"Length initial data: {len(data)}")
+        logging.info(f"Rows dropped with NA's: {initial_length - dropped_length}")
+        logging.info(f"Length filtered data: {len(processed_data)}\n")
+
+        return processed_data
+
+    except FileNotFoundError as e:
+        logging.error(f"Error: File not found. Details {e}")
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:129]: Why are you capturing this exception? Which function call can raise this, you are not opening any file

File: src/module_4/push_exploration_non_linear.md [hunk 5]
```diff
@@ -0,0 +1,2005 @@
...
+        logging.info(f"Date from: {start_date}")
+        logging.info(f"Date to: {end_date}")
+
+        # Based on the number of days, we get the train days
+        num_days = len(daily_orders)
+        train_days = num_days - test_days - validation_days
+
+        # Train
+        train_start = daily_orders.index.min()
+        train_end = train_start + timedelta(days=train_days)
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:194]: Nit pick, you are using `train_days+1` instead of `train_days`: if `train_start` is the 5 and `train_days=2`, then `train_end = 5+2 = 7`

File: src/module_4/push_exploration_non_linear.md [hunk 6]
```diff
@@ -0,0 +1,2005 @@
...
+
+        # Validation (no need to define test)
+        validation_end = train_end + timedelta(days=validation_days)
+
+        if cols is None:
+            # Use all columns if cols is not specified
+            cols = data.columns.tolist()
+
+        train = data[data.order_date <= train_end][cols]
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:203]: very elegant!

File: src/module_4/push_exploration_non_linear.md [hunk 7]
```diff
@@ -0,0 +1,2005 @@
...
+        logging.info(f"Test set ratio: {len(test) / len(data):.2%}")
+
+        X_train, y_train = split_sets(train, label)
+        X_val, y_val = split_sets(val, label)
+        X_test, y_test = split_sets(test, label)
+
+        return X_train, y_train, X_val, y_val, X_test, y_test
+
+    # not the best practice to catch all exceptio
+    except KeyError as ke:
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:220]: any reason to be capturing `KeyError` again if you are already capturing at the beginning of the function

File: src/module_4/push_exploration_non_linear.md [hunk 8]
```diff
@@ -0,0 +1,2005 @@
...
+    - y_test (array-like): True labels.
+    - save_curves_path (str, optional): Directory to save the generated figure.
+    If None, the figure will not be saved.
+
+    Returns:
+    - None
+    """
+
+    # Create a timestamp for unique filenames
+    timestamp = datetime.now().strftime("%Y_%m_%d")
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:244]: I suggest to put any kind of file name generation in the top (outside any function) of the file, as global variable

File: src/module_4/push_exploration_non_linear.md [hunk 9]
```diff
@@ -0,0 +1,2005 @@
...
+# Non-linear models
+
+We will apply different models and compare against previous results, in the following order:
+- Decision Trees
+- Random Forests
+- Gradient Bosting Trees
+- XGBoost
+- Catboost
+
+In that sense, let's modify the generate_evaluation_curves to plot multiple funcs at the same time, in order to compare visually the performance of each model vs. baseline.
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:432]: good idea!

File: src/module_4/push_exploration_non_linear.md [hunk 10]
```diff
@@ -0,0 +1,2005 @@
...
+    and display them in a single figure.
+    """
+
+    # Create a timestamp for unique filenames
+    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
+
+    plt.figure(figsize=(10, 5))
+
+    # Plot for each model
+    for model_name, y_pred in zip(model_names, y_preds):
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:472]: nit pick: I suggest to keep 2 functions: the original one for the AUC for a single prediction and the one which goes through the different predictions with the for, get the metrics for each of them and joins them

File: src/module_4/push_exploration_non_linear.md [hunk 11]
```diff
@@ -0,0 +1,2005 @@
...
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_14_0.png)
+    
+
+
+Obviously, a single decision tree will hardly beat it. So we will try using random forests with different params of n_trees. As as curiosity, if we take a look at the plot for LR with same parameters as trained before, the AUC for PR is 0.16 vs 0.17 before (C, solver and  regularisation remains equal). I think this might happen as model regularisation can still have some randomness in model weights, hence getting different results. For addressing that we could have made use of random_seed.
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:545]: This is not obvious to me -> a single decision tree should in general be more expressive than LR (i.e. reduce its bias), however they overfit like crazy (i.e. increase its variance) (and that's my bet of what's going on here). So the way to confirm this would be first to plot the same curves for the training set. Decision trees have their own methods (nor very sophisticated but exists) to overcome overfitting: mainly branch prunning and minimal cost-complexity prunning. I'm pretty sure if you tune them properly you can achieve a better or similar performance than LR

File: src/module_4/push_exploration_non_linear.md [hunk 12]
```diff
@@ -0,0 +1,2005 @@
...
+
+## Random Forests
+
+
+```python
+# Train Logistic Regression
+lr = LogisticRegression(penalty="l2", C=1e-5, solver="liblinear")
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:552]: ML Important homework: you are using l2 penalty (ridge) to overcome overfitting, but you are never checking if that's the case: for that compare performance in train and test. Play with the C parameter to see how overfitting is reduced and bias is increased

File: src/module_4/push_exploration_non_linear.md [hunk 13]
```diff
@@ -0,0 +1,2005 @@
...
+generate_evaluation_curves_multi(model_names, y_preds, y_val)
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_18_0.png)
+    
+
+
+For ROC we see no improvement, whereas higher number of trees works better than just a few. This model didn't improve our baseline defined by the linear model in blue.
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:581]: This is defo a little bit weird (that RF are not superior to LR), maybe it'd be interesting to play a little bit more with it. specially tune the parameters of the trees and look what happens in train vs test

File: src/module_4/push_exploration_non_linear.md [hunk 14]
```diff
@@ -0,0 +1,2005 @@
...
+
+In Precision-Recall, we obtain similar AUC for 200 trees, with better results for RandomForest between 0-15% recall, and performing worse than logistic regression beyond 20% recall (we see blue and brown lines intersecting at that point). This would be interesting if we knew in advance the exact point we would like to be in, which we know more or less but will need to make more adjustments.
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:583]: Also, an important learning here: even if I'm quite sure that RF can finally beat LR, you already can see that it won't be by too much! In general LR works very well both as a baseline and a lot of times even as the final model in production

File: src/module_4/push_exploration_non_linear.md [hunk 15]
```diff
@@ -0,0 +1,2005 @@
...
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_18_0.png)
+    
+
+
+For ROC we see no improvement, whereas higher number of trees works better than just a few. This model didn't improve our baseline defined by the linear model in blue.
+
+In Precision-Recall, we obtain similar AUC for 200 trees, with better results for RandomForest between 0-15% recall, and performing worse than logistic regression beyond 20% recall (we see blue and brown lines intersecting at that point). This would be interesting if we knew in advance the exact point we would like to be in, which we know more or less but will need to make more adjustments. 
+
+As a result, no model has yet defeated our beatiful logistic regression. It is also worth to mention that training for this n_trees took around 20min, while lr run x10 faster - in case scalability and deployment plays an important role.
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:585]: It is also worth to mention that training for this n_trees took around 20min
some comment about scalability: RF is an algorithm which can be paralelized very easily (just fit each tree in parallel). Also, by controlling the parameter `max_features` you can reduce a lot its computation power. Furthermore, there are some more sophisticated improvements like the ones applied in lightgbm (which also apply to normal tries) which make everything much faster. So fast/scalability is quite relative here

File: src/module_4/push_exploration_non_linear.md [hunk 16]
```diff
@@ -0,0 +1,2005 @@
...
+If performance could be a significant issue in this case, we can consider discarding from 'count_pets' to the right at a later time. However, we would continue in this case with all the cols
+
+## Gradient Boosting Trees
+
+
+```python
+from sklearn.ensemble import GradientBoostingClassifier
+
+# Let's be conservative for reducing total time
+n_trees_values = [5, 100, 250]
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:657]: This doesn't make much sense: bear in mind that in order to train with 250 trees, you first need to train with 1, then with 2, so 5 and 100 will be included too :P I just recommend to train with as many trees as you want to plot the learning curve of how does you metric evolve with the number of trees

File: src/module_4/push_exploration_non_linear.md [hunk 17]
```diff
@@ -0,0 +1,2005 @@
...
+max_depth_values = [1, 3, 5]
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:658]: I'd say depth 1 is too small

Comment (gsantam) [reply src/module_4/push_exploration_non_linear.md:658]: Also bear in mind that the most important parameter of GBM is the learning rate (same for NN), so feel free to play with it

File: src/module_4/push_exploration_non_linear.md [hunk 18]
```diff
@@ -0,0 +1,2005 @@
...
+
+        trained_gbt_models.append(gbt)
+        y_preds_gbt.append(gbt.predict_proba(X_val)[:, 1])
+        gbt_model_names.append(f"GBT (n_trees={n_trees}, depth={depth})")
+```
+
+
+```python
+lr_model_name = ["LogisticRegression"]
+rf_model_names = [f"RandomForest (n_trees={n})" for n in n_trees_values]
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:680]: Mistake?

File: src/module_4/push_exploration_non_linear.md [hunk 19]
```diff
@@ -0,0 +1,2005 @@
...
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_53_0.png)
+    
+
+
+Indeed they do, my gut-feeling it is that we are providing more complexity to the model, hence letting it decide with more robustness for positive outcomes (although they are not our best model as we saw previously).
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:1452]: Good observation and nice you went deeper into this. I think your gut feeling is not very far away from reality. To understand what's going on, bear in mind that to emit a probability in a certain leaf, the model is looking into all the samples falling into that leaf and looking at the proportion of them which are 1s. If we have a single tree with no branches and one top leaf, the emited probability should be the proportion of 1s in the whole dataset (which is small). As we add more branches (or more trees) the model starts "overfitting" and grouping a larger proportion of ones into the same leaf, increasing the emited probabilities for them

**Review by gsantam — CHANGES_REQUESTED** [review]

File: src/module_4/push_exploration_non_linear.md [hunk 20]
```diff
@@ -0,0 +1,2005 @@
...
+)
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_26_0.png)
+    
+
+
+GBT with n_trees = 100 and depth = 3 seems to be our best model (brown line, legend is wrong for depth values but do not want to recompute bc of runtime) yet very similar to LR. As additional info, all models took around 30 min to run (again, if we need that info into account).
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:699]: > all models took around 30 min to run
Again, I'm pretty sure that can be improved

File: src/module_4/push_exploration_non_linear.md [hunk 21]
```diff
@@ -0,0 +1,2005 @@
...
+from sklearn.preprocessing import OneHotEncoder
+
+def calibrate_model(model, X_train, y_train, method='sigmoid'):
+    # Fit the calibrator - note that we do not need a Pool object here
+    calibrated = CalibratedClassifierCV(model, method=method, cv='prefit')
+    calibrated.fit(X_train, y_train)
+    return calibrated
+
+# Assuming X_val is your validation feature matrix and y_val are the validation labels:
+calibrated_catboost = calibrate_model(catboost_model, X_val, y_val)
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:1468]: you should always calibrate the model in train! otherwise you can do overfitting without noticing (does the calibration generalizes to other sets)

File: src/module_4/push_exploration_non_linear.md [hunk 22]
```diff
@@ -0,0 +1,2005 @@
...
+        X_train, y_train = split_sets(train, label)
+        X_val, y_val = split_sets(val, label)
+        X_test, y_test = split_sets(test, label)
+
+        return X_train, y_train, X_val, y_val, X_test, y_test
+
+    # not the best practice to catch all exceptio
+    except KeyError as ke:
+        logging.error(f"Key Error: {str(ke)}")
+        return
```
Comment (gsantam) [inline src/module_4/push_exploration_non_linear.md:222]: I thought you would do a moving train / val moving window in order to evaluate the model over different periods but I don't see you do this

**Review by gsantam — CHANGES_REQUESTED** [review]

File: src/module_4/model_fit.py [hunk 23]
```diff
@@ -0,0 +1,460 @@
...
+
+    Parameters:
+    - event (dict): Dictionary with input parameters such as path, params, etc.
+
+    Returns:
+    - dict: Dictionary with model information such as model name, path, etc.
+    """
+    try:
+        # Extract parameters from event or set defaults
+        data_path = event.get("data_path")
```
Comment (gsantam) [inline src/module_4/model_fit.py:373]: ideally you should have these in a config file: https://docs.python.org/3/library/configparser.html#:~:text=A%20configuration%20file%20consists%20of,removed%20from%20keys%20and%20values.

**Review by gsantam — CHANGES_REQUESTED** [review]
Added few comments, nothing too severe, mostly about playing a little bit more with the models and always comparing the performance in train and test to see if a model has a bias or variance problem
